trace/trace.go

```diff
@@ -93,39 +152,69 @@ func (t *Trace) logWithStepThreshold(stepThreshold time.Duration) {
 		writeFields(&buffer, t.fields)
 		buffer.WriteString(" ")
 	}
-	buffer.WriteString(fmt.Sprintf("(started: %v) (total time: %v):\n", t.startTime, totalTime))
-	lastStepTime := t.startTime
-	for _, step := range t.steps {
-		stepDuration := step.stepTime.Sub(lastStepTime)
-		if stepThreshold == 0 || stepDuration > stepThreshold || klog.V(4).Enabled() {
-			buffer.WriteString(fmt.Sprintf("Trace[%d]: [%v] [%v] ", tracenum, step.stepTime.Sub(t.startTime), stepDuration))
-			buffer.WriteString(step.msg)
-			if len(step.fields) > 0 {
-				buffer.WriteString(" ")
-				writeFields(&buffer, step.fields)
-			}
-			buffer.WriteString("\n")
-		}
-		lastStepTime = step.stepTime
-	}
+	buffer.WriteString(fmt.Sprintf("(%v) (total time: %vms):", t.startTime.Format("02-Jan-2006 15:04:00.000"), totalTime.Milliseconds()))
+	lastStepTime := writeTrace(&buffer, t, fmt.Sprintf("\nTrace[%d]: ", tracenum), stepThreshold)
 	stepDuration := endTime.Sub(lastStepTime)
 	if stepThreshold == 0 || stepDuration > stepThreshold || klog.V(4).Enabled() {
-		buffer.WriteString(fmt.Sprintf("Trace[%d]: [%v] [%v] END\n", tracenum, endTime.Sub(t.startTime), stepDuration))
+		buffer.WriteString(fmt.Sprintf("\nTrace[%d]: [%v] [%v] END\n", tracenum, endTime.Sub(t.startTime), stepDuration))
 	}
 
 	klog.Info(buffer.String())
 }
 
+func writeTrace(b *bytes.Buffer, t *Trace, formatter string, stepThreshold time.Duration) time.Time {
+	lastStepTime := t.startTime
+	stepAndTraces := t.stepsTraces
+	if len(stepAndTraces) == 0 {
+		return lastStepTime
+	}
+	for _, stepOrTrace := range stepAndTraces {
+		stepTime := stepOrTrace.writeStep(b, formatter, t, stepThreshold, lastStepTime)
+		lastStepTime = stepTime
+	}
+	return lastStepTime
+}
+
 // LogIfLong is used to dump steps that took longer than its share
 func (t *Trace) LogIfLong(threshold time.Duration) {
-	if time.Since(t.startTime) >= threshold {
+	t.threshold = &threshold
+	if time.Since(t.startTime) >= threshold || t.parentTrace == nil {
 		// if any step took more than it's share of the total allowed time, it deserves a higher log level
-		stepThreshold := threshold / time.Duration(len(t.steps)+1)
+		stepThreshold := calculateStepThreshold(t)
 		t.logWithStepThreshold(stepThreshold)
+	} else {
+		for _, s := range t.stepsTraces {
+			nestedTrace, ok := s.(*Trace)
+			if ok && nestedTrace.threshold != nil && time.Since(nestedTrace.startTime) >= *nestedTrace.threshold {
+				stepThreshold := calculateStepThreshold(nestedTrace)
+				nestedTrace.logWithStepThreshold(stepThreshold)
```

Comment:
Does this just goes one level deep in the nested traces? I was expecting any nested trace that needs to be logged would get logged directly if none of its parents get logged, no matter how deeply nested. Good thing to have a test case for.

Reply:
The function is recursive so it should work. I will write some tests for it.

Reply:
Okay, I was expecting the for loop surrounding this call to logWithStepThreshold to itself be in logWithStepThreshold for a recursive approach, but maybe I don't follow how the recursion works.

Reply:
Oh true. I missed that. Thank you